### ISSUE TITLE
Service with session affinity cannot isolate the fault appropriately

### ISSUE THREAD

**Issue by luozhiwenn** [kubernetes/kubernetes#110302]
### What happened?

By setting service.spec.sessionAffinity to "ClientIP", connections from a particular client are passed to the same Pod each time when accessing a Service.
When the Pod that the client continuously accesses becomes not ready, the client will still access the NotReady Pod because of the iptables rules, even if the maximum session sticky time has been exceeded, which caused the request of the client cannot be passed to the ready Pod.

### What did you expect to happen?

Service with session affinity can isolate the fault appropriately, after the maximum session sticky time, traffic can be redirected to one of the ready Pods.

### How can we reproduce it (as minimally and precisely as possible)?

1. kube-proxy runnnig with iptables proxy mode
2. create a session affinity Service with two endpoints
3. make a client accesses the Service periodically
4. stop one of the endpoints(the client accesses) and make it not ready
5. the client  accesses the notready endpoint periodically

### Anything else we need to know?

kube-proxy runnnig with iptables proxy mode in my cluster.

### Kubernetes version

<details>

```console
$ kubectl version
v1.22.1
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by thockin** [comment]
In this case it sounds like your client has an open connection, right (as opposed to closing it and periodically re-opening)?  Or am I misunderstanding?  If so, this is about connection tracking, not affinity.

Session affinity only matters for new connections, not new packets on existing connections.  In (almost?) every implementation, kube-proxy is more of a router than a proxy, which means your client pod is connected to the server directly, with some address translation in between.  

https://github.com/kubernetes/kubernetes/issues/108523#issuecomment-1071832988

I think that it is correct to leave open sessions to not-ready endpoints - this is important for gracefully draining traffic when things like deployment updates are happening.

If you need your clients to disconnect, you have to do that yourself from within the server.

**Comment by luozhiwenn** [comment]
> In this case it sounds like your client has an open connection, right (as opposed to closing it and periodically re-opening)? Or am I misunderstanding? If so, this is about connection tracking, not affinity.

Sorry for not clearly described, what I mean is `closing it and periodically re-opening`.

Maybe kube-proxy shouldn't renew iptables recent record when endpoints become not ready? This will cause the fault endpoints cannot be isolate appropriately: affected by the affinity rules, new connections on this node will proxy to not-ready endpoint(assume that the client on this node has connected to this backend.).

I think it is necessary to determine whether the endpoint is ready in the following logic.
https://github.com/kubernetes/kubernetes/blob/bd2776e0c930ab149d8c0656e83f979fb017ce0a/pkg/proxy/iptables/proxier.go#L1019

@thockin

**Comment by aojea** [comment]
> I think it is necessary to determine whether the endpoint is ready in the following logic.

Kube-proxy doesn't renew iptables rules for you unready endpoint, actually it delete those rules

**Comment by luozhiwenn** [comment]
> Kube-proxy doesn't renew iptables rules for you unready endpoint, actually it delete those rules

mm, the issue seems to have been fixed by [#106497](https://github.com/kubernetes/kubernetes/pull/106497)
sorry~I'll take some time to reconfirm

**Comment by aojea** [comment]
> > Kube-proxy doesn't renew iptables rules for you unready endpoint, actually it delete those rules
> 
> mm, the issue seems to have been fixed by [#106497](https://github.com/kubernetes/kubernetes/pull/106497) sorry~I'll take some time to reconfirm

/close

thanks for letting us know, can you close the PR too?

**Comment by luozhiwenn** [comment]
> thanks for letting us know, can you close the PR too?

closed

**Comment by aojea** [comment]
by the way, I was checking the old code and already checks for readiness

https://github.com/kubernetes/kubernetes/blob/e3eac9677a785ef5e66f5af787bbeed260ed6040/pkg/proxy/iptables/proxier.go#L1030-L1051

there is a `continue` that skips not ready endpoints before the line you are pointing ... do we have a bug in old versions?

**Comment by luozhiwenn** [comment]
> there is a continue that skips not ready endpoints before the line you are pointing ... do we have a bug in old versions?

Thanks for your reminding. I investigated again and found that the problem was actually fixed by [#106030](https://github.com/kubernetes/kubernetes/pull/106030):
![image](https://user-images.githubusercontent.com/74550097/204070691-3ef01b66-0d40-4ee4-b95f-20c965006400.png)
